redact.go

```diff
@@ -1,186 +1,323 @@
-		new = strings.ReplaceAll(new, ")", "")
-		if _, err := strconv.Atoi(new); err == nil {
-			*items = append(*items, temp)
+func matchDOB(input string) (matches []match) {
+	var start int
+	var length int
+	var isCandidate bool
+	var monthStart int
+	var monthLength int
+	var monthCandidate bool
+
+	for i := 0; i < len(input)-1; i++ {
+		character := input[i]
+		if _, found := used[i]; found {
+			continue
+		}
+		if !isNumeric(character) {
+			if breakNotFound(character){
+				monthLength++
+				start = i + 1
+				length = 0
+				isCandidate = false
+				continue
+			}
+			if isMonth(input[monthStart:monthStart + monthLength]){
+				monthCandidate = true
+				continue
+			}
+			monthStart = i + 1
+			monthLength = 0
+			if isDOB(length) {
+				matches = appendMatches(matches, start, length)
+				length = 0
+				start = i + 1
+				continue
+			}
+		}
+		if isCandidate || monthCandidate {
+			length++
+		} else {
+			isCandidate = true
+			start = i + 1
+		}
+		if length == 2 && monthCandidate {
+			matches = appendMatches(matches, monthStart, monthLength + length + 1)
+			monthCandidate = false
+			length = 0
+			start = 0
+			monthStart = 0
+			monthLength = 0
 		}
 	}
-}
-func appendTelCandidate(temp string, items *[]string, min, max int) {
-	lengthTemp := len(temp)
-	if lengthTemp == 11 && strings.Contains(temp, "-") {
-		return
+	if isNumeric(input[len(input)-1]) {
+		length++
 	}
-	if resemblesCC(temp, lengthTemp) {
-		return
+	if isDOB(length) {
+		matches = appendMatches(matches, start, length)
 	}
-	appendCandidate(temp, items, min, max)
+	return matches
 }
-func resemblesCC(temp string, lengthTemp int) bool {
-	return lengthTemp > 12 && !((strings.Contains(temp, "(") || strings.Contains(temp, ")")) || strings.Contains(temp, " ") || strings.Contains(temp, "-"))
+
+func isDOB(length int) bool {
+	return length >= 6 && length <= 10
 }
-func spaceDelimitedCandidate(input string, i int) bool {
-	return i+1 < len(input) && ('0' <= input[i+1] && input[i+1] <= '9') && ('0' <= input[i-1] && input[i-1] <= '9' || input[i-1] == ')')
+func isMonth(month string) bool {
+	_, found := months[month]
+	return found
+}
+func isNumeric(value byte) bool {
+	return value >= '0' && value <= '9'
+}
+
+func breakNotFound(character byte) bool {
+	return character != '-' && character != ' ' && character != '.' && character != '(' && character != ')' && character != '/'
+}
+func appendMatches(matches []match, start, length int) []match {
+	for i := start; i <= start+length; i++ {
+		used[i] = struct{}{}
+	}
+	return append(matches, match{InputIndex: start, Length: length})
+}
+type match struct {
+	InputIndex int
+	Length     int
 }
 
-var (
-	datesWithDashes  = regexp.MustCompile(`\d{1,2}-\d{1,2}-\d{4}`)
-	datesWithSlashes = regexp.MustCompile(`\d{1,2}/\d{1,2}/\d{4}`)
-	datesInUSFormat  = regexp.MustCompile(`[a-zA-Z]{3,9} \d{1,2}, \d{4}`)
-	datesInEUFormat  = regexp.MustCompile(`\d{1,2} [a-zA-Z]{3,9} \d{4}`)
+var(
+	used = make(map[int]struct{})
```

Comment:
What is the purpose of this package-level variable?

Reply:
This allows us to designate which index values have already been accounted for to be changed to asterisks. There may be a more efficient way to do this but the issue that we were facing was the phone number was picking up numbers that were actually credit cards, this was causing issues with spacing as it was adding additional asterisks where they did not belong. Any thoughts on a more effective way to do this?

Reply:
Ok, makes sense.